File: 113.path-sum-ii.js
/*
 * @lc app=leetcode id=113 lang=javascript
 *
 * [113] Path Sum II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function (root, targetSum) {
  if (!root) return [];
  if (!root.left && !root.right && root.val === targetSum) {
    return [[root.val]];
  }

  const subRes1 = pathSum(root.left, targetSum - root.val);
  const subRes2 = pathSum(root.right, targetSum - root.val);

  const res = [];
  subRes1.forEach((list) => {
    res.push([root.val, ...list]);
  });
  subRes2.forEach((list) => {
    res.push([root.val, ...list]);
  });
  return res;
};
// @lc code=end

File: solutions/113.path-sum-ii.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} target_sum
# @return {Integer[][]}
def path_sum(root, target_sum)
  ans = []
  tmp = []

  dfs = lambda do |node, target|
    return if node.nil?
    ans.push tmp.dup << node.val if node.val == target && node.left.nil? && node.right.nil?
    tmp.push node.val
    dfs.call(node.left, target - node.val)
    dfs.call(node.right, target - node.val)
    tmp.pop
  end

  dfs.call root, target_sum
  ans
end

# alternative
def path_sum(root, target_sum)
  return [] if root.nil?
  return [[root.val]] if root.val == target_sum && root.left.nil? && root.right.nil?

  path_sum(root.left, target_sum-root.val).map { [root.val] + _1 } + path_sum(root.right, target_sum-root.val).map { [root.val] + _1 }
end
